coralproject/talk
# Asset Rewriting

## What does this PR solve?

A situation came up during testing that highlighted an interesting situation. On some sites, it may be possible to navigate to it without SSL (via HTTP, not HTTPS), which under some circumstances may have been completely an unintentional side effect of not enabling HSTS headers.

To resolve the situation where the URL's need to change, or there is duplicate comment streams, the best recommendation is to use a combination of CLI tools to rectify the situation.

- You can use the `./bin/cli assets list` to find assets in the system.
- You can use the `./bin/cli assets merge <srcID> <dstID>` to merge two assets by their ID's, moving all comments into the sources ID.
- **You can now use `./bin/cli assets rewrite <searchPattern> <replacementPattern>` that will use a regex search + replace of urls.** This can be extremely useful if you are switching to a new domain, or changing your site from http/https to https-only.

The best method to mitigate all of these is to provide a asset specific ID into the embed using the:

```js
Talk.render(el, {
  // ...
  asset_id: "<myAssetID>",
  // ...
});
```

And the https://github.com/coralproject/talk-plugin-custom-asset-id plugin. (See the github page for the plugin to learn about installation)

## What does this PR do?

- Adds the new `./bin/cli assets rewrite <searchPattern> <replacementPattern>` command.

## How do you test it?

1. Visit http://localhost:3000/assets/title/One
2. Create a single comment of "This is on page One"
3. Go to your CLI and run `./bin/cli assets rewrite "^(.*)/One" "\$1/Two" --dry-run`
4. See the output changing the url from http://localhost:3000/assets/title/One to http://localhost:3000/assets/title/Two
5. Run the CLI without the `--dry-run` flag: `./bin/cli assets rewrite "^(.*)/One" "\$1/Two"`
6. Visit http://localhost:3000/assets/title/Two
7. Notice your comment of "This is on page One" now exists on page "/Two"!
8. Visit http://localhost:3000/assets/title/One again, and notice no comments at all!

Tracking: https://www.pivotaltracker.com/story/show/154224188
